### ISSUE TITLE
Fully-fledged support for mocked objects

### ISSUE THREAD

**Issue by ServerlessArchitect** [python/mypy#9776]
**Feature**

I'd like to propose that fully-fledged support for mocked objects be added so that type checks work with spec'd mocks.

```
import mock
import typing


class Foo:
    def bar(self, a: int) -> str:
        return str(a)


# I want to pass a spec'd mock here (mock.create_autospec(spec=Foo)).
def test_something(foo: Foo):
    # I'd like to find a way for this to be accepted by mypy.
    # error: "Callable[[int], str]" has no attribute "return_value"
    foo.bar.return_value = "42"


# This can be resolved arduously by writing a custom protocol for each mock.
class BarMockProtocol(typing.Protocol):
    return_value: str

    def __call__(self, a: int) -> str:
        pass

    # def assert_called():
    #     ...
    # And so on...


class FooMockProtocol(typing.Protocol):
    bar: BarMockProtocol


def test_something_else(foo: FooMockProtocol):
    # Error as expected.
    # error: Incompatible types in assignment (expression has type "int", variable has type "str")
    foo.bar.return_value = 42

    # No error as expected.
    foo.bar.return_value = "42"


# This is what I'd suggest
def test_something_else_yet(foo: mock.Mock[Foo]):
    # Error
    foo.bar.return_value = 42
    # No error
    foo.bar.return_value = "42"


# Usage sample
foo_mock = mock.create_autospec(spec=Foo)
test_something_else_yet(foo_mock)
```

**Pitch**

Currently, it's difficult to fully utilise mypy with mocks. I can see the following options.
1. Use `typing.Any` to disable type checks.
2. Use the type of the mocked class and live/deal with errors when using mock-specific attributes and methods (e.g. `return_value`, `assert_*`).
3. Write a custom protocol for every mocked class as demonstrated above.

In other words, it's a choice between boilerplate and limitations. Please do let me know if I missed anything.

**Comment by hauntsaninja** [comment]
mock.create_autospec is currently stubbed as returning `Any`, so your options 2 and 3 might not even always work. I'd probably just go with your option 1 and type code that relies on your mock being a mock with Any as well.
You'd probably need intersection types to make this kind of thing work in general, and that's not going to happen soon.

**Comment by ServerlessArchitect** [comment]
@hauntsaninja Thank you for your swift response! Intersection types indeed sound like the solution. In the meantime, do you think this might be solved with a mypy plugin? I wouldn't mind giving one a go but also wouldn't want to set off on something that's doomed to fail.

Another possibility would be to find a way to globally ignore the errors related to `return_value`, `assert_*`, and so on. I had two ideas in mind:
1. Tell mypy not to lint statements that refer to those specific attributes, e.g. through a regex exclusion rule. This way, I'd still be able to capture genuine usage errors.
1. Coerce mypy into thinking that those specific properties are always valid as if they were built-in.

Such configuration would, of course, be applied to my test directory only. I appreciate those might sound sort of dubious, but considering they'd only apply to the tests, I'd be pretty happy with the trade-offs.

Do you think either of the above is something that may become reality? If so, I can go ahead and create a new issue.

**Comment by ServerlessArchitect** [comment]
@hauntsaninja I hope you've been well. I wanted to check if the above made sense. Thank you!

**Comment by hauntsaninja** [comment]
I'm actually not too familiar with mypy plugins, so my answer is maybe! I'd definitely look into it, maybe someone on Gitter would be around to help. You might need to use custom stubs for mock to get things to work.
Another option for you to explore is using custom protocols, but use multiple inheritance so it's not so tedious.

I don't think mypy would implement your attribute suggestions; they feel pretty ad hoc to me (you're of course welcome to open issues in case others feel different). A workaround for you here is to write a script that calls mypy with --show-error-codes, then perform your own filtering based on the error code `[attr-defined]` and the attribute name mentioned.

**Comment by ServerlessArchitect** [comment]
@hauntsaninja The script a great idea. We'll go ahead this way for now, hoping that eventually mocks can become first-class citizens :)

**Comment by Bolognafingers** [comment]
Encountered this problem today, are there any plans on supporting this in a near future?

**Comment by antoniogamiz** [comment]
It would be nice to have this.

**Comment by dhensen** [comment]
Anybody knows any blogpost that explains how to work around this issue?
